ssl/statem/statem_clnt.c

```diff
@@ -821,7 +821,10 @@ int tls_construct_client_hello(SSL *s)
         goto err;
     }
 
-    if ((sess == NULL) || (sess->ssl_version != s->version) ||
+    if ((sess == NULL) ||
+        /* Only destroy session which version is bigger than maximum supported.
+         */
+        ssl_version_cmp(s, sess->ssl_version, s->version) > 0 ||
```

Comment:
I'm talking about the version locked methods. So, for example, if someone has used TLSv1_1_client_method() (instead of the version flexible TLS_client_method()) but the session version is TLSv1.0. According to this logic, that would be ok, but actually it's doomed to fail.